Add inline color pickers to script editor

dugramen opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
![image](https://github.com/user-attachments/assets/df781876-ce61-4e6c-9c00-4aedc89a01ad)
-->
Implements https://github.com/godotengine/godot-proposals/issues/6653

![image](https://github.com/user-attachments/assets/3cc31723-b547-40d4-9813-590b2f817dea)


The proposal already had a linked PR which added color buttons to the gutter. 
This PR instead adds color buttons directly in the text area, and supports multiple colors per line, like in VSCode. Had to make some `text_server_adv` fixes to support objects with 0 span length.

The following constructors are supported:
- `Color()` (Empty, an html color string, or 3-4 rgba values)
- `Color.from_hsv()`
- `Color.from_rgba8()`
- `Color.from_ok_hsl`

The mode / format of the constructor will match which mode is selected in the color picker. Video demo:

https://github.com/user-attachments/assets/1d9fec3c-d881-448c-8d79-aee22bae64b4

Commit:
Add color pickers to script editor

JekSun97 commented:
I think we should have a corresponding setting in the editor settings to turn this display mode on and off

bruvzg commented on scene/gui/text_edit.cpp:
```diff
@@ -1562,6 +1609,12 @@ void TextEdit::_notification(int p_what) {
 					}
+
+					HashMap<int, Variant> glyph_to_obj;
+					for (Variant k : TS->shaped_text_get_objects(rid)) {
+						glyph_to_obj[TS->shaped_text_get_object_glyph(rid, k)] = k;
+					}
```
You can use `shaped_text_get_object_rect` and draw all of these at once, without mapping it to the glyphs.

Something like:
```c++
for (Variant k : TS->shaped_text_get_objects(rid)) {
	Rect2 col_rect = TS->shaped_text_get_object_rect(rid, k);
	col_rect.position += Vector2(char_margin + ofs_x, ofs_y);
	if (!clipped && (col_rect.position.x + char_margin) >= xmargin_beg && (col_rect.position.x + col_rect.size.x + char_margin) <= xmargin_end) {
		Array arr = k;
		if (arr.size() == 4 && arr[3].get_type() == Variant::Type::COLOR) {
			draw_rect(col_rect, Color(arr[3]));
		}
		draw_rect(col_rect, Color(1, 1, 1), false, 1);
	}
}
```

bruvzg reviewed (dismissed):
Also, `0` size spans should not exist and can properly work, at least one replacement character is needed for BiDi.

bruvzg commented:
> Also, 0 size spans should not exist and can properly work, at least one replacement character is needed for BiDi.

Actually it's probably fine, but a note should be added to the documentation to mention that if it's `0` size it is attached to the previous character.

bruvzg commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -6114,6 +6117,11 @@ bool TextServerAdvanced::_shaped_text_update_breaks(const RID &p_shaped) {
 						}
 					}
+					// Do not add extra space for color picker object.
+					if (sd_glyphs[i].start == sd_glyphs[i].end || (i + 1 < sd->glyphs.size() && sd_glyphs[i + 1].start == sd_glyphs[i + 1].end)) {
```
```suggestion
					if (((sd_glyphs[i].flags & GRAPHEME_IS_EMBEDDED_OBJECT) == GRAPHEME_IS_EMBEDDED_OBJECT && sd_glyphs[i].start == sd_glyphs[i].end) || (i + 1 < sd->glyphs.size() && (sd_glyphs[i + 1].flags & GRAPHEME_IS_EMBEDDED_OBJECT) == GRAPHEME_IS_EMBEDDED_OBJECT && sd_glyphs[i + 1].start == sd_glyphs[i + 1].end)) {
```

This condition is too generic and should not apply to anything except embedded objects.

dugramen replied:
Ok. The same check is in like 1 or 2 other places. I'll update them there too

dugramen commented:
Ok I applied all the changes by @bruvzg, and added an editor setting to enable / disable this. Also rebased just in case

bruvzg commented on scene/gui/text_edit.cpp:
```diff
@@ -2284,6 +2346,44 @@ void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
+									inline_color_start = i_start;
+									inline_color_end = i_end;
+
+									Vector2 xf_mpos = mb->get_global_position();
```
```suggestion
									Vector2 xf_mpos = mb->get_position() + get_screen_position();
```

Or it will be off if the editor window is not maximized/not on the primary screen.

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -1216,6 +1216,9 @@
 			The shape of the caret to use in the script editor. [b]Line[/b] displays a vertical line to the left of the current character, whereas [b]Block[/b] displays an outline over the current character.
 		</member>
+		<member name="text_editor/appearance/enable_inline_color_picker" type="bool" setter="" getter="">
+			If [code]true[/code], colored buttons will appear before color constructors. Clicking them will popup a color picker for visually editing the color's parameters.
```
A link to the **Color** Variant type is important, I feel.
```suggestion
			If [code]true[/code], displays a colored button before any [Color] constructor in the script editor. Clicking on them allows the color to be modified through a color picker.
```
Perhaps **ColorPicker**, too, but not as much, and it does not parse right for localization.

Mickeon reviewed (commented):
From what I know, this looks to be the first feature that makes it so that what is displayed in the **TextEdit** does not exactly match its underlying text.
This could open the door for more personalized displays, such as parameter names, akin to other IDEs. There was a proposal about this but I cannot find it at the moment. 

Either way, I think this is a bit too niche to be exposed to the scripting API, and it's a good thing it's not in this PR...
Isn't the code for this too invasive? It affects **TextEdit** directly for no justifiable reason, I feel.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -353,10 +356,18 @@ class TextEdit : public Control {
 	PopupMenu *menu = nullptr;
 	PopupMenu *menu_dir = nullptr;
 	PopupMenu *menu_ctl = nullptr;
+	PopupPanel *color_popup = nullptr;
+	ColorPicker *color_picker = nullptr;
```
Is it possible to move the ColorPicker to ScriptTextEditor? The TextEdit doesn't use it on its own.

dugramen commented:
> It affects **TextEdit** directly for no justifiable reason, I feel.

Well all TextServer access happens in TextEdit, so objects need to be added in there. But I get what you mean that its weird for color picker code to be in TextEdit. Maybe I can generalize it a bit and move the color related stuff into CodeEdit, and have TextEdit only call it through Callables

> Is it possible to move the ColorPicker to ScriptTextEditor? The TextEdit doesn't use it on its own.

Yeah I'll move color picker to that or CodeEdit too

dugramen commented:
Ok I generalized everything in TextEdit, so it can be used for any inline objects with span length 0. All color related functionality was moved to ScriptTextEditor. I added enums and validation functions to make it easier to understand & build off of in the future. Also applied the suggestions.

(I'm not sure why the tests are failing. That line was working fine before and I haven't changed anything in color_picker.h or color_picker.cpp)

kitbdev commented:
It looks like in #105892 ColorPicker's `_text_type_toggled` was put in tools only ifdefs, so you should wrap the methods you added to ColorPicker in `#ifdef TOOLS_ENABLED` too.

dugramen commented:
Ok that worked. Thanks!

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -396,8 +396,215 @@ bool ScriptTextEditor::show_members_overview() {
+	code_editor->get_text_editor()->end_complex_operation();
+}
+
+void ScriptTextEditor::_picker_opening() {
+	previous_tooltip_enabled = code_editor->get_text_editor()->is_symbol_tooltip_on_hover_enabled();
+	code_editor->get_text_editor()->set_symbol_tooltip_on_hover_enabled(false);
+}
```
If this is called when the picker is already open, it gets called before `_picker_closed` and causes the symbol tooltip on hover to stay off.

kitbdev reviewed (commented):
- Supersedes https://github.com/godotengine/godot/pull/76171

Ideally the color parsing and text replacing would be separated to GDScript ScriptLanguage or somewhere, but there are lots of other GDScript specific things in the script editor, so it's fine for now.

Parsing should make sure that it only gets color constructors and not functions that end with `Color`:

![image](https://github.com/user-attachments/assets/e060a7d7-1f9a-42a2-8bc6-4617f9b26ca3)

The mouse cursor should change to indicate the color box can be clicked.

Selecting color objects should probably show the selection background behind it.

KoBeWi commented:
There is another syntax supported for Colors, but it's probably the least used - integer colors, e.g. `Color(0xff0000ff)`.

Mickeon commented:
> There is another syntax supported for Colors, but it's probably the least used - integer colors, e.g. `Color(0xff0000ff)`.

Still, if not supported it may be deemed to be a bug down the line.

Mickeon commented on scene/gui/text_edit.h:
```diff
@@ -121,6 +121,14 @@ class TextEdit : public Control {
 	};
 
+	/* Inline Info. */
+	enum InlineInfo {
+		INFO_LINE,
+		INFO_COLUMN,
+		INFO_WIDTH_RATIO, // Used to determine the object's width given a line height.
+		INFO_MAX
+	};
```
It's not odd. If you think about it this is akin to the constants used for PopupMenu items.

kitbdev replied:
PopupMenu constants are used to distinguish which action to do based on the button, so only one callback is needed, but here it's used to transfer data generically in an Array with various types.

Like this:
```cpp
Array color_info;
color_info.resize(INFO_COLOR_MAX);
color_info[TextEdit::INFO_LINE] = p_line;
color_info[TextEdit::INFO_COLUMN] = i_start;
color_info[TextEdit::INFO_WIDTH_RATIO] = 1.0;
color_info[INFO_COLOR_END] = i_par_end;

color_info[INFO_COLOR] = Color(color_string);
```

So it's already basically being used like a dictionary.

Mickeon replied:
Does it even have to be either one of the two? Can it not be a struct here?

dugramen replied:
It has to be a variant to be added as a object to TextServer. A struct can't do that (as far as I know)

Mickeon replied:
In my opinion using Arrays here is okay, as a strictly "server-sided" operation. It occupies less memory and doesn't need to be dynamic. It reminds me of the way ArrayMesh works.

kitbdev commented:
> > There is another syntax supported for Colors, but it's probably the least used - integer colors, e.g. `Color(0xff0000ff)`.
> 
> Still, if not supported it may be deemed to be a bug down the line.

There's more constructors that aren't supported, I thought it was intentional. `Color.hex`, `Color.hex64`, `Color.html`, `Color.from_rgb9995`, `Color(code, alpha)`.
Do they all need to be supported in this PR? Since they aren't as common I figured they could be added later, if wanted.

dugramen commented:
I have an updated commit ready for all kitbdev's fixes, but my wifi is out so I'll have to push tomorrow or something

> There is another syntax supported for Colors, but it's probably the least used - integer colors, e.g. Color(0xff0000ff).

The way I have it set up now, the used syntax is just whatever tab is selected in the color picker, and if the hex toggle is enabled it uses a hex string. But that limits it to just those 4 constructors.

Would it be better to just ignore the tab, and instead repurpose the hex button to cycle between all the constructors. I think vscode does that

dugramen commented:
Alright wifi came back. Summary of fixes:
- Switched to using dictionaries (willing to switch back if arrays are better for performance)
- Fixed drawing bugs, and added transparent checker image for translucent colors
- Prevent functions that end with "color" from showing a picker
- Draw selection background for objects
- Changed cursor to pointer when hovering an object

> If this is called when the picker is already open, it gets called before _picker_closed and causes the symbol tooltip on hover to stay off.

It seems that ScriptTextEditor already forces hover tooltip to enabled anyways (not based on some setting like I thought), so I got rid of `previous_tooltip_enabled` and just enabled / disabled it when the popup opens / closes

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -1432,6 +1483,19 @@ void TextEdit::_notification(int p_what) {
 						char_margin += wrap_indent;
 					}
 
+					// Validate inline objects.
+					Vector<Dictionary> object_keys;
+					if (inline_object_drawer.is_valid()) {
+						for (int l = 0; l < ldata->get_line_count(); l++) {
+							for (Variant k : ldata->get_line_objects(l)) {
+								if (!is_inline_info_valid(k)) {
+									continue;
+								}
+								object_keys.push_back(k);
+							}
+						}
+					}
```
This is in the loop that goes over each line wrap, so it shouldn't need to loop over line count. The current one is `line_wrap_index`.

dugramen replied:
Oh this explains a lot. I was wondering why there were multiple lines per line data. Completely forgot about word wrap

kitbdev reviewed (commented):
Previous issues are fixed.

When using word wrap and a line with a color object is wrapped, I get this error:
`ERROR: modules\text_server_adv\text_server_adv.cpp:7137 - Condition "!sd->objects.has(p_key)" is true. Returning: Rect2()`

dugramen commented:
Ok new commit implementing the fixes, and added support for hex integers. 

I changed it so that you now choose a constructor from a dropdown, rather than it being tied to which tab the color picker is on. This enables any number of constructors in the future. I also simplified the code a bit to make adding & reordering constructors easier. Here's a demo:

https://github.com/user-attachments/assets/95d64146-2143-4519-ba20-c2fbdb8ea116

Also after testing more while considering word wrap, I found a few bugs relating to object hover / click position, so had to rework that a bit. The weird indentation offset code is just how its calculated in other places too

kitbdev reviewed (approved):
Works well.

dugramen commented on editor/plugins/script_text_editor.h:
```diff
@@ -160,6 +169,16 @@ class ScriptTextEditor : public ScriptEditorBase {
+		MODE_RGB,
+		MODE_HSV,
+		MODE_OKHSL,
+		MODE_RGB8,
```
Delayed this a bit but I just pushed with the updated order. Also made the option list's font the same as the code editor, for readability

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -2292,6 +2372,37 @@ void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
+						}
+						Dictionary info = k;
+						Rect2 obj_rect = ldata->get_line_object_rect(wrap_i, k);
+						obj_rect.position.x += xmargin_beg + wrap_indent;
```
There is an issue when hovering/clicking when scrolled horizontally.
I think `first_visible_col` needs to be used in the calculation.

dugramen replied:
Ok updated it

Repiteo commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -6122,6 +6127,11 @@ bool TextServerAdvanced::_shaped_text_update_breaks(const RID &p_shaped) {
 						}
 					}
+					// Do not add extra space for color picker object.
+					if (((sd_glyphs[i].flags & GRAPHEME_IS_EMBEDDED_OBJECT) == GRAPHEME_IS_EMBEDDED_OBJECT && sd_glyphs[i].start == sd_glyphs[i].end) || (i + 1 < sd->glyphs.size() && (sd_glyphs[i + 1].flags & GRAPHEME_IS_EMBEDDED_OBJECT) == GRAPHEME_IS_EMBEDDED_OBJECT && sd_glyphs[i + 1].start == sd_glyphs[i + 1].end)) {
```
MSVC compilation failed for me when compiling, citing a `<` signed/unsigned mismatch. Could be solved by either making `i` a `uint32_t`, or fixing in-place:
```suggestion
					if (((sd_glyphs[i].flags & GRAPHEME_IS_EMBEDDED_OBJECT) == GRAPHEME_IS_EMBEDDED_OBJECT && sd_glyphs[i].start == sd_glyphs[i].end) || (uint32_t(i + 1) < sd->glyphs.size() && (sd_glyphs[i + 1].flags & GRAPHEME_IS_EMBEDDED_OBJECT) == GRAPHEME_IS_EMBEDDED_OBJECT && sd_glyphs[i + 1].start == sd_glyphs[i + 1].end)) {
```

dugramen replied:
Ok updated it

Repiteo commented:
Thanks!

PIZZA-ALERT commented:
Thanks for adding this. Sorry if this has been asked, but can you please also update it to work for cases where the color's word is typed? Like instead of Color(0,0,0), I may type Color.BLACK, and currently, this doesn't support the color picker button.
<img width="142" height="30" alt="image" src="https://github.com/user-attachments/assets/1c33dd85-637c-4f4e-9156-574cbd76b359" />
<img width="96" height="24" alt="image" src="https://github.com/user-attachments/assets/8937df88-59fc-4677-8a4b-0acc678e400c" />

dugramen commented:
@PIZZA-ALERT kitbdev has a PR for it https://github.com/godotengine/godot/pull/109104
